README.md
# flureejs-tx


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
(modification: no type change headlines) and this project adheres to
[Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.1.3] - 2020-09-07

### Changed

- Facilities for getting cmd and signature
- Documentation and test improvements

## [0.1.2] - 2020-08-30

### Changed

- Working Karma into Chrome and Firefox
- Documentation and test improvements

## [0.1.1] - 2020-08-23

### Changed

- Changes for use flureejs-config package
- Documentation and test improvements

## [0.1.0] - 2020-08-21

- Basic functionality

[0.1.3]: https://github.com/StylusFrost/flureejs-tx/compare/v0.1.2...v0.1.3
[0.1.2]: https://github.com/StylusFrost/flureejs-tx/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/StylusFrost/flureejs-tx/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/StylusFrost/flureejs-tx/releases/v0.1.0


test/types.ts
// The type of ./txs.json
export type TxsJsonEntry = {
  privateKey: string
  sendersAuthID: string
  type: string
  raw: string[]
}


test/index.spec.ts
/* tslint:disable no-invalid-this */
import * as assert from 'assert'
import Transaction from '../src'
import { TxsJsonEntry } from './types'
import { zeros, privateToPublic, toBuffer } from 'flureejs-utils'
const txFixtures: TxsJsonEntry[] = require('./txs.json')
describe('[Transaction]: Basic functions', function() {
  const txs: Transaction[] = []
  it('should decode txs', function() {
    txFixtures.slice(0, 1).forEach(function(tx: any) {
      const pt = new Transaction(tx.raw)
      assert.equal('0x' + pt.type.toString('hex'), tx.raw[0])
      assert.equal('0x' + pt.db.toString('hex'), tx.raw[1])
      assert.equal('0x' + pt.tx.toString('hex'), tx.raw[2])
      assert.equal('0x' + pt.auth.toString('hex'), tx.raw[3])
      assert.equal('0x' + pt.fuel.toString('hex'), tx.raw[4])
      assert.equal('0x' + pt.nonce.toString('hex'), tx.raw[5])
      assert.equal('0x' + pt.expire.toString('hex'), tx.raw[6])
      assert.equal('0x' + pt.v.toString('hex'), tx.raw[7])
      assert.equal('0x' + pt.r.toString('hex'), tx.raw[8])
      assert.equal('0x' + pt.s.toString('hex'), tx.raw[9])
      txs.push(pt)
    })
  })

  it('should verify Signatures', function() {
    txs.forEach(function(tx) {
      assert.equal(tx.verifySignature(), true)
    })
  })

  it('should not verify Signatures', function() {
    txs.forEach(function(tx) {
      tx.s = zeros(32)
      assert.equal(tx.verifySignature(), false)
    })
  })

  it('should give a string about not verifing Signatures', function() {
    txs.forEach(function(tx) {
      assert.equal(tx.validate(true).slice(0, 17), 'Invalid Signature')
    })
  })

  it('should validate', function() {
    txs.forEach(function(tx) {
      assert.equal(tx.validate(), false)
    })
  })

  it('should sign tx', function() {
    txs.forEach(function(tx, i) {
      if (txFixtures[i].privateKey) {
        const privKey = new Buffer(txFixtures[i].privateKey, 'hex')
        tx.sign(privKey)
      }
    })
  })
  it("should get sender's AuthID after signing it", function() {
    txs.forEach(function(tx, i) {
      if (txFixtures[i].privateKey) {
        assert.equal(tx.getSenderAuthID(), txFixtures[i].sendersAuthID)
      }
    })
  })

  it("should get sender's public key after signing it", function() {
    txs.forEach(function(tx, i) {
      if (txFixtures[i].privateKey) {
        assert.equal(
          tx.getSenderPublicKey().toString('hex'),
          privateToPublic(new Buffer(txFixtures[i].privateKey, 'hex')).toString('hex'),
        )
      }
    })
  })

  it("should get sender's authID after signing it (second call should be cached)", function() {
    txs.forEach(function(tx, i) {
      if (txFixtures[i].privateKey) {
        assert.equal(tx.getSenderAuthID(), txFixtures[i].sendersAuthID)
        assert.equal(tx.getSenderAuthID(), txFixtures[i].sendersAuthID)
      }
    })
  })

  it('should verify signing it', function() {
    txs.forEach(function(tx, i) {
      if (txFixtures[i].privateKey) {
        assert.equal(tx.verifySignature(), true)
      }
    })
  })

  it('should accept lesser r values', function() {
    const tx = new Transaction()
    tx.r = toBuffer('0x0005')
    assert.equal(tx.r.toString('hex'), '05')
  })
})


src/types.ts
/**
 * Any object that can be transformed into a `Buffer`
 */
export interface TransformableToBuffer {
  toBuffer(): Buffer
}

/**
 * A hex string prefixed with `0x`.
 */
export type PrefixedHexString = string

/**
 * A Buffer, hex string prefixed with `0x`, Number, or an object with a toBuffer method such as BN.
 */
export type BufferLike = Buffer | TransformableToBuffer | PrefixedHexString | number

export interface TxData {
  /**
   * The transaction's type.
   */
  type?: BufferLike

  /**
   * The transaction's db to.
   */
  db?: BufferLike

  /**
   * The body of the transaction
   */
  tx?: BufferLike

  /**
   * Authid is signing
   */

  auth?: BufferLike

  /**
   * 	Max integer for the amount of fuel to use for this transaction
   */

  fuel?: BufferLike

  /**
   * 	Nonce, to ensure that the command map is unique.
   */
  nonce?: BufferLike

  /**
   * 	Epoch milliseconds after which point this transaction can no longer be submitted.
   */
  expire?: BufferLike

  /**
   * EC recovery ID.
   */
  v?: BufferLike

  /**
   * EC signature parameter.
   */
  r?: BufferLike

  /**
   * EC signature parameter.
   */
  s?: BufferLike
}


src/index.ts
import {
  ecsign,
  publicToAuthID,
  ecrecover,
  toBuffer,
  defineProperties,
  bufferToInt,
  BN,
} from 'flureejs-utils'
import { TxData } from './types'

export const enum TxType {
  TX = 'tx',
}

export default class Transaction {
  public raw!: Buffer[]
  public type!: Buffer
  public db!: Buffer
  public tx!: Buffer
  public auth!: Buffer
  public fuel!: Buffer
  public nonce!: Buffer
  public expire!: Buffer
  public v!: Buffer
  public r!: Buffer
  public s!: Buffer

  protected _from?: Buffer
  private _senderPubKey?: Buffer

  /**
   * Creates a new transaction from an object with its fields' values.
   *
   * @param data - A tx can be initialized with object containing them by name.
   *
   * @example
   * ```js
   * const txData = {
   *   type: '0x7478',
   *   db: '0x61756469747a6f6e652d746573746e65742f7365616c6462',
   *   tx: '0x5b7b225f6964223a225f636f6c6c656374696f6e247365616c222c226e616d65223a227365616c222c22646f63223a224120636f6c6c656374696f6e20746f20686f6c642074686520696e666f726d6174696f6e206f6620746865207365616c73227d5d',
   *   auth: '0x5466386f764864676e445a58724d7a71454c706131787331636664684a696533507761',
   *   fuel: '0x31303030303030',
   *   nonce: '0x36',
   *   expire: '0x31353937393236363939373339',
   *   v: '0x1c',
   *   r: '0x5e1d3a76fbf824220eafc8c79ad578ad2b67d01b0c2425eb1f1347e8f50882ab',
   *   s: '0x5bd428537f05f9830e93792f90ea6a3e2d1ee84952dd96edbae9f658f831ab13'
   * };
   * const q = new Transaction(txData);
   * ```
   */

  constructor(data: TxData = {}) {
    // Define Properties
    const fields = [
      {
        name: 'type',
        allowZero: false,
        default: new Buffer([]),
      },
      {
        name: 'db',
        allowZero: false,
        default: new Buffer([]),
      },
      {
        name: 'tx',
        allowZero: false,
        default: new Buffer([]),
      },
      {
        name: 'auth',
        allowZero: false,
        default: new Buffer([]),
      },
      {
        name: 'fuel',
        length: 32,
        allowZero: false,
        allowLess: true,
        default: new Buffer([]),
      },
      {
        name: 'nonce',
        length: 32,
        allowZero: false,
        allowLess: true,
        default: new Buffer([]),
      },
      {
        name: 'expire',
        length: 32,
        allowZero: false,
        allowLess: true,
        default: new Buffer([]),
      },
      {
        name: 'v',
        allowZero: true,
        default: new Buffer([]),
      },
      {
        name: 'r',
        length: 32,
        allowZero: true,
        allowLess: true,
        default: new Buffer([]),
      },
      {
        name: 's',
        length: 32,
        allowZero: true,
        allowLess: true,
        default: new Buffer([]),
      },
    ]

    // attached serialize
    defineProperties(this, fields, data)

    /**
     * @property {Buffer} from (read only) sender address of this transaction, mathematically derived from other parameters.
     * @name from
     * @memberof Transaction
     */
    Object.defineProperty(this, 'from', {
      enumerable: true,
      configurable: true,
      get: this.getSenderAuthID.bind(this),
    })

    this._validateV(this.v)
    this._overrideVSetterWithValidation()
  }

  private _validateV(v?: Buffer): void {
    if (v === undefined || v.length === 0) {
      return
    }

    const vInt = bufferToInt(v)

    if (vInt === 27 || vInt === 28) {
      return
    }
  }

  private _overrideVSetterWithValidation() {
    const vDescriptor = Object.getOwnPropertyDescriptor(this, 'v')!

    Object.defineProperty(this, 'v', {
      ...vDescriptor,
      set: v => {
        if (v !== undefined) {
          this._validateV(toBuffer(v))
        }
        vDescriptor.set!(v)
      },
    })
  }

  /**
   * returns the sender's authID
   */
  getSenderAuthID(): Buffer {
    if (this._from) {
      return this._from
    }
    const pubkey = this.getSenderPublicKey()
    this._from = publicToAuthID(pubkey)
    return this._from
  }

  /**
   * returns the public key of the sender
   */
  getSenderPublicKey(): Buffer {
    if (!this.verifySignature()) {
      throw new Error('Invalid Signature')
    }

    // If the signature was verified successfully the _senderPubKey field is defined
    return this._senderPubKey!
  }

  /**
   * Determines if the signature is valid
   */
  verifySignature(): boolean {
    try {
      const v = bufferToInt(this.v)
      this._senderPubKey = ecrecover(this.cmd(), v, this.r, this.s)
    } catch (e) {
      return false
    }
    return !!this._senderPubKey
  }

  /**
   * Validates the signature and checks to see if it has enough gas.
   */
  validate(): boolean
  validate(stringError: false): boolean
  validate(stringError: true): string
  validate(stringError: boolean = false): boolean | string {
    const errors = []
    if (!this.verifySignature()) {
      errors.push('Invalid Signature')
    }

    if (stringError === false) {
      return errors.length === 0
    } else {
      return errors.join(' ')
    }
  }

  /**
   * sign a transaction with a given private key
   * @param privateKey
   */

  sign(privateKey: Buffer) {
    const sig = ecsign(this.cmd(), privateKey)
    Object.assign(this, sig)
  }
  public cmd(): Buffer {
    return Buffer.from(
      JSON.stringify({
        type: this.type.toString(),
        db: this.db.toString(),
        tx: JSON.parse(this.tx.toString()),
        auth: this.auth.toString(),
        fuel: new BN(this.fuel, 'hex').toNumber(),
        nonce: new BN(this.nonce, 'hex').toNumber(),
        expire: new BN(this.expire, 'hex').toNumber(),
      }),
    )
  }

  public signature(): Buffer {
    const newR = this.r[0] & 0x80 ? Buffer.concat([hexToUnit8Array('00'), this.r]) : this.r
    const newS = this.s[0] & 0x80 ? Buffer.concat([hexToUnit8Array('00'), this.s]) : this.s
    const result =
      '02' +
      newR.length.toString(16) +
      newR.toString('hex') +
      '02' +
      newS.length.toString(16) +
      newS.toString('hex')

    return Buffer.from(
      this.v.toString('hex') + '30' + Buffer.from(result, 'hex').length.toString(16) + result,
      'hex',
    )

    function hexToUnit8Array(str: string) {
      return new Uint8Array(Buffer.from(str, 'hex'))
    }
  }
}
